### PR TITLE
Fix: type extension

### LINKED ISSUE #776: Possible edge case
Trying to use ocamlformat on a large project, I've started getting this error message:

```
ocamlformat: Cannot process "redacted.ml".
  Please report this bug at https://github.com/ocaml-ppx/ocamlformat/issues.
  BUG: ast changed.
```

I've whittled the problem down to a rather specific-looking case:

```
let%any_extension f = function()
  | _ -> ()
```

Basically, any time a top-level `let`-extension defines a `function`. Am I doing something obviously wrong?

### PR DESCRIPTION
**Pull request by gpetiot**
Fix #776

### CODE REVIEW

File: src/Fmt_ast.ml [hunk 1]
```diff
@@ -3668,10 +3671,16 @@ and fmt_structure_item c ~last:last_item ?ext {ctx; ast= si} =
       let fmt_decl ~first ~last decl =
         let pre =
           if first then
-            if Poly.(rec_flag = Recursive) then "type " else "type nonrec "
+            let pre =
+              if Poly.(rec_flag = Recursive) then "type" else "type nonrec"
+            in
+            if Option.is_some ext then pre else pre ^ " "
```
Comment (Julow) [inline src/Fmt_ast.ml]: Minor comment:
This line is a bit inelegant. The `?pre` argument of `fmt_type_declaration` never needs to be `None` and is always terminated by a space character. (Except for `with` constraints but I get no regression) What do you think of adding the space in `fmt_type_declaration`, in every cases?

Comment (gpetiot) [reply src/Fmt_ast.ml]: Done.

**Comment by gpetiot** [comment]
> On signatures I get:
> 
> ```diff
> -type%foo t = < .. >
> +[%%foo: type t = < .. >]
> ```
> 
> There is the exact same code for types in signatures.

I refactored the code and added a test. test_branch looks good.

File: test/passing/js_source.ml.ref [hunk 2]
```diff
@@ -3548,8 +3548,10 @@ end
 module rec Typ : sig
   module type PAIR = sig
     type t
- and t1
- and t2
+
+    and t1
+
+    and t2
```
Comment (Julow) [inline test/passing/js_source.ml.ref]: This is a fix.

**Pull request merged**